### ISSUE TITLE
CoreWebView2InitializationCompleted fires before IsInitialized (private prop) would return true when using Fixed Version run time

### ISSUE THREAD

**Issue by dqwork** [MicrosoftEdge/WebView2Feedback#1050]
**Description**
This happened to me when using the Fixed Version binaries and its more obiovus it happens, because when a second corewebview is initialised it doesn't have the custom folder specified, how the root cause happens regardless.

When initialising the CoreWebView2 run time. The CoreWebView2InitializationCompleted event fire's before IsInitiliazed (private prop of WebView2 class) would return true, because _initTask's (private field) status is WaitingForActivation not RanToCompletion.

This means that when, after waiting for the core webview to be initialised, I to navigate to a starting url (Which I can only do in or after that event since I need to tell it where to find the run time so can't rely on auto initialistion), by setting the Source property - The Source setter checks IsInitialized (which we know is false), if it is false, recreates the runtime or attempts to anyway but never fires the CoreWebView2InitializationCompleted event again . This leads to the page not rendering. Also the settings I applied in the event handler like 'Don't show ContextMenus' are ignored.

**Version**
SDK: 1.0.705.50, 1.0.774.44 (tried both)
Runtime: 89.0.774.48
Framework: Win Forms .Net Framework 4.62
OS: Windows 10

**Repro Steps**
I've pasted a copy of the main form code from a very basic reproduciable example below. 

```
using Microsoft.Web.WebView2.Core;
using System;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Windows.Forms;

namespace WebView2_Testing
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
            webView.CoreWebView2InitializationCompleted += WebView_CoreWebView2Ready;
            InitializeAsync();
        }

        private void WebView_CoreWebView2Ready(object sender, EventArgs e)
        {
            var initTaskField = webView.GetType().GetField("_initTask", BindingFlags.Instance | BindingFlags.NonPublic);
            dynamic initTask = initTaskField.GetValue(webView);
            Debug.WriteLine($"WebView._initTask - Status: {initTask?.Status ?? "<null>"}");
            webView.CoreWebView2.Settings.AreDefaultContextMenusEnabled = false;
            webView.CoreWebView2.Settings.AreDevToolsEnabled = false;
            webView.CoreWebView2.Settings.IsZoomControlEnabled = false;
            webView.Source = new Uri("https://www.google.com");
        }

        private async void InitializeAsync()
        {
            await webView.EnsureCoreWebView2Async(null);
        }

        private void refreshBtn_Click(object sender, EventArgs e)
        {
            webView.Reload();
        }
    }
}
```

**Additional context**
There is currently a work around for this. Don't use the source property, directly call the webView.CoreWebView2.Navigate method which doesn't cause another CoreWebView2 to be created. But ideally I'd like to not have to have this work around and a long comment to explain why in my code base :) . Another work around is not to set the source in the event handler, do it after the awaited EnsureCoreWebView2Async completes

[AB#32079191](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/32079191)

**Comment by cgeier** [comment]
Try the following: 

```
using System;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Diagnostics;
using System.Reflection;
using Microsoft.Web.WebView2.Core;
using Microsoft.Web.WebView2.WinForms;
using System.IO;

namespace WebView2_Testing
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private async void Form1_Load(object sender, EventArgs e)
        {
            webView.CoreWebView2InitializationCompleted += WebView_CoreWebView2InitializationCompleted;

            LogMsg("before InitializeAsync");
            await InitializeAsync(webView);
            LogMsg("after InitializeAsync");

            //set properties
            webView.CoreWebView2.Settings.AreDefaultContextMenusEnabled = false;
            webView.CoreWebView2.Settings.AreDevToolsEnabled = false;
            webView.CoreWebView2.Settings.IsZoomControlEnabled = false;

            webView.CoreWebView2.Navigate("https://www.google.com");
            LogMsg("after Navigate");
        }

      
        public async Task InitializeAsync(WebView2 wv, string webCacheDir = "")
        {
            CoreWebView2EnvironmentOptions options = null;
            string tempWebCacheDir = string.Empty;
            CoreWebView2Environment webView2Environment = null;

            //set value
            tempWebCacheDir = webCacheDir;

            if (String.IsNullOrEmpty(tempWebCacheDir))
            {
                //get fully-qualified path to user's temp folder
                tempWebCacheDir = System.IO.Path.GetTempPath();

                tempWebCacheDir = System.IO.Path.Combine(tempWebCacheDir, System.Guid.NewGuid().ToString("N"));
            }

            //webView2Environment = await CoreWebView2Environment.CreateAsync(@"C:\Program Files (x86)\Microsoft\Edge Dev\Application\89.0.774.48", tempWebCacheDir, options);
            //webView2Environment = await CoreWebView2Environment.CreateAsync(null, tempWebCacheDir, options);

            //WebView2 Fixed Version Runtime
            webView2Environment = await CoreWebView2Environment.CreateAsync("Microsoft.WebView2.FixedVersionRuntime.89.0.774.48.x86", tempWebCacheDir, options);


            LogMsg("Info: before EnsureCoreWebView2Async");

            //wait for CoreWebView2 initialization
            await wv.EnsureCoreWebView2Async(webView2Environment);

            LogMsg("Info: after EnsureCoreWebView2Aync");

            LogMsg(wv.CoreWebView2.Environment.BrowserVersionString);

            LogMsg("Info: Cache data folder set to: " + tempWebCacheDir);

            var initTaskField = webView.GetType().GetField("_initTask", BindingFlags.Instance | BindingFlags.NonPublic);
            dynamic initTask = initTaskField.GetValue(wv);

            LogMsg($"WebView._initTask - Status: {initTask?.Status ?? "<null>"}");

            
        }

        private void LogMsg(string msg)
        {
            string logMsg = String.Format("{0} {1}", DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss:fff"), msg);
            System.Diagnostics.Debug.WriteLine(logMsg);
        }

        private void refreshBtn_Click(object sender, EventArgs e)
        {
            //webView.Visible = false;
            //webView.Dispose();
            webView.Reload();
        }


        private void WebView_CoreWebView2InitializationCompleted(object sender, Microsoft.Web.WebView2.Core.CoreWebView2InitializationCompletedEventArgs e)
        {
            LogMsg("WebView_CoreWebView2InitializationCompleted");
        }
    }
}
```

**Comment by dqwork** [comment]
@cgeier Thanks for taking the time to help out. I will most certainly give that a go. But would you agree that your comment is a work around, and there is an issue?

Looking at the microsoft official '[Getting Started](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.webbrowser.allownavigation?view=net-5.0)' page. In section 8 it first starts to mention the async nature of initialising the webview.

```
public Form1()
{
    InitializeComponent();
    this.Resize += new System.EventHandler(this.Form_Resize);
    webView.NavigationStarting += EnsureHttps;
    InitializeAsync();
}

async void InitializeAsync()
{
    await webView.EnsureCoreWebView2Async(null);
}
```

My example does the same and I do understand the mechanics of the fact I can't await in a ctor, that's exactly why I wait for the event 'CoreWebView2InitializationCompleted' - its just that it appears that when this event is fired... the initialization isn't completed, which to me seems like a bug.

Or have I got the wrong of the stick?

**Comment by champnic** [comment]
Thanks a lot for the detailed investigation @dqwork - I've added this bug to our backlog to get fixed.

**Comment by starburst997** [comment]
I'm nearly sure I got this same bug using the Evergreen version (so it is not specific to "Fixed Version"). 

Sometimes it will takes multiple seconds to load (made the app look very cheap), I was doing some debugging using `NavigateToString` since it was inconsistent when I noticed the `IsInitialized` of the `Source` property. Calling `Navigate` directly instead and I haven't encountered this issue.

**Comment by mrsteamfist** [comment]
@starburst997 Just to let you know, NavigateToString will not work until the WebView has been manually initialized.
https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.wpf.webview2?view=webview2-dotnet-1.0.774.44

- Call the EnsureCoreWebView2Async(CoreWebView2Environment) method. This is referred to as explicit initialization.
-Set the Source property (which could be done from markup, for example). This is referred to as implicit initialization. Either option will start initialization in the background and return back to the caller without waiting for it to finish. To specify options regarding the initialization process, either pass your own CoreWebView2Environment to EnsureCoreWebView2Async(CoreWebView2Environment) or set the control's CreationProperties property prior to initialization.

**Comment by champnic** [comment]
This is fixed in 1.0.865-prerelease SDK package. Thanks!
